File: core/java/src/main/java/hara/lang/protocol/IHash.java
package hara.lang.protocol;

// import hara.lang.base.G;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.ihash", name = "IHash")
public interface IHash {

  default long hashCalc() {
    // return hashCalc(hashType());
    return hashCalc(Constant.HashType.RAPID);
  }

  long hashCalc(Constant.HashType t);

  @HaraMethod(value = "hash", arity = 1)
  default long hashGet() {
    // return hashCalc(hashType());
    return hashCalc(Constant.HashType.RAPID);
  }

  default long hashGet(Constant.HashType t) {
    return hashCalc(t);
  }

  default String hashSeed() {
    return "";
  }

  default Constant.HashType hashType() {
    // return G.DEFAULT_HASH;
    return Constant.HashType.RAPID;
  }
}

File: core/rust/src/lang/protocol/ihash.rs
use hara_protocol_macros::hara_protocol;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HashType {
    System,
    Rapid,
    Murmur3,
    Sip,
}

#[hara_protocol(namespace = "std.protocol.ihash", name = "IHash")]
pub trait IHash {
    fn hash_calc(&self, hash_type: HashType) -> u64;

    #[hara_method(value = "hash", arity = 1)]
    fn hash(&self) -> u64 {
        self.hash_calc(self.hash_type())
    }

    fn hash_get(&self) -> u64 {
        self.hash()
    }

    fn hash_get_as(&self, hash_type: HashType) -> u64 {
        self.hash_calc(hash_type)
    }

    fn hash_type(&self) -> HashType {
        HashType::Rapid
    }
}
